File: src/main/java/rag/trace/TraceBus.java
package rag.trace;
import java.util.*;


public class TraceBus {

    private List<TraceSink> sinks = new ArrayList<>();

    // register a new TraceSink (Observer) to receive trace events
    public void addSink(TraceSink sink) { sinks.add(sink); }
    public void publish(TraceEvent ev) throws Exception {
        for (TraceSink sink : sinks) {
            try {
                sink.record(ev);
            } catch (RuntimeException ex) {
                // print the error message
                System.err.printf(
                    "Error recording event in sink %s: %s%n", 
                    sink.getClass().getSimpleName(), 
                    ex.getMessage()
                );
            }
        }
    }

    // Optionally provides read-only access to the list of sinks for monitoring/debugging.
    public List<TraceSink> getSinks() { return this.sinks; }
}

File: src/main/python/rag/trace/trace_bus.py
from typing import List
from rag.trace.trace_sink import TraceSink
from rag.trace.trace_event import TraceEvent


class TraceBus:
    def __init__(self) -> None:
        self.sinks: List[TraceSink] = []

    def add_sink(self, sink: TraceSink) -> None:
        self.sinks.append(sink)

    def publish(self, e: TraceEvent) -> None:
        for sink in self.sinks:
            try:
                sink.accept(e)
            except RuntimeError as ex:
                print(f"Trace sink failed: {str(ex)}", file=__import__('sys').stderr)






